### PR TITLE
G9rXf4Mm Enable admins to change user's email

### PR DESCRIPTION
**Pull request by j-maynard**
As per the prototype, user managers and GDS users should be able to change the users' emails.  This should fulfil that criteria.

### CODE REVIEW

File: app/views/users/show_update_email.erb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<h1 class="govuk-heading-xl">Change email</h1>
```
Comment (jakubmiarka) [inline app/views/users/show_update_email.erb]: Localize please

File: app/views/users/show_update_email_verification_code.erb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+<div class="govuk-grid-row">
```
Comment (jakubmiarka) [inline app/views/users/show_update_email_verification_code.erb]: What is this view for?

File: spec/controllers/users_controller_spec.rb [hunk 3]
```diff
@@ -95,6 +95,23 @@
...
+    describe '#show_update_email' do
+      it 'renders the update email page' do
+        get :show_update_email, :params => { :user_id => user_id }
+
+        expect(response).to have_http_status(:success)
+        expect(subject).to render_template(:show_update_email)
+      end
+    end
+    
+    describe '#update_email' do
```
Comment (jakubmiarka) [inline spec/controllers/users_controller_spec.rb]: There are 4 logical branches in that controller, therefore I would expect at least 4 tests + 1 where `update_user_email_form` param is missing.

File: app/views/users/show_update_email.erb [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+<%= form_for @form, url: update_user_email_address_post_path, class: 'govuk-form-group' do |f| %>
+  <div class="govuk-form-group">
+    <%= f.label :email, class: "govuk-label" %>
+    <%= f.email_field :email, autofocus: true, autocomplete: "email", class: "govuk-input govuk-input--width-20" %>
+  </div>
+
+  <div class="actions">
+    <%= f.submit t('users.show_update_email.button'), class: "govuk-button", data: { module: "govuk-button" } %>
+  </div>
+<% end %>
```
Comment (jakubmiarka) [inline app/views/users/show_update_email.erb]: I would expect the file extension to be `html.erb`

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -71,6 +71,27 @@ def new
...
 
+  def show_update_email
+    @form = UpdateUserEmailForm.new({})
+  end
+
+  def update_email
+    @form = UpdateUserEmailForm.new(params["update_user_email_form"])
+    if @form.valid?
+      update_user_email(user_id: params["user_id"], email: @form.email)
+      redirect_to update_user_path
```
Comment (jakubmiarka) [inline app/controllers/users_controller.rb]: does this work? I would expect it to be `update_user_path(user_id: params[:user_id])`

File: Gemfile.lock [hunk 6]
```diff
@@ -49,7 +49,7 @@ GEM
     arel (9.0.0)
     ast (2.4.0)
     aws-eventstream (1.0.3)
-    aws-partitions (1.220.0)
+    aws-partitions (1.221.0)
```
Comment (jakubmiarka) [inline Gemfile.lock]: Is this related to this PR/change?

File: lib/auth/authentication_backend.rb [hunk 7]
```diff
@@ -181,6 +182,21 @@ def get_user(user_id:)
...
+      user_pool_id: user_pool_id,
+      username: user_id,
+      user_attributes: [
+        { name: "email", value: email },
+        { name: 'email_verified', value: 'True' },
+      ],
+    )
+  rescue Aws::CognitoIdentityProvider::Errors::AliasExistsException => e
+    raise AliasExistsException.new(e)
+  rescue StandardError => e
```
Comment (jakubmiarka) [inline lib/auth/authentication_backend.rb]: would prefer if we were rescuing the AWS ServiceError exception rather than a generic StandardError

File: app/views/users/show_update_email.erb [hunk 8]
```diff
@@ -0,0 +1,12 @@
+<h1 class="govuk-heading-xl">Change email</h1>
+
+<%= form_for @form, url: update_user_email_address_post_path, class: 'govuk-form-group' do |f| %>
```
Comment (jakubmiarka) [inline app/views/users/show_update_email.erb]: If the field is invalid it should use the design system error pattern

File: app/models/update_user_email_form.rb [hunk 9]
```diff
@@ -0,0 +1,11 @@
+class UpdateUserEmailForm
+  include ActiveModel::Model
+
+  attr_reader :email
+  validates_presence_of :email
+  validates :email, format: { with: URI::MailTo::EMAIL_REGEXP }
```
Comment (jakubmiarka) [inline app/models/update_user_email_form.rb]: you should be able to just do `validates :email, email: true`

File: app/controllers/users_controller.rb (hunk 5 above)
Comment (j-maynard) [reply app/controllers/users_controller.rb]: We're not very consistent about this in the users_controller or anywhere else for that matter.  It seems you can use either a symbol or a string when it comes to the params coming in to a controller.

Comment (jakubmiarka) [reply app/controllers/users_controller.rb]: I meant the `redirect_to` line. Rails supplies the parameter automatically, which is nice. No changes needed then...

**Comment by jakubmiarka** [comment]
closing in favour of https://github.com/alphagov/verify-self-service/pull/234
